app/gui2/src/components/ComponentBrowser.vue

```diff
@@ -19,52 +37,301 @@ function positionAtMouse(): boolean {
   return true
 }
 
-const transform = computed(() => {
-  const nav = props.navigator
-  const pos = scenePosition.value
-  return `${nav.transform} translate(${pos.x}px, ${pos.y}px) scale(${
-    1 / nav.scale
-  }) translateY(-100%)`
+// === Components List and Positions ===
+
+const componentStore = useComponentsStore()
+
+const visibleComponents = computed(() => {
+  if (scroller.value) {
+    const first_visible = componentAtY(animatedScrollPosition.value)
+    const last_visible = componentAtY(animatedScrollPosition.value + scrollerSize.value.y)
+    return componentStore.components.slice(first_visible, last_visible + 1).map((component, i) => {
+      return { component, index: i + first_visible }
+    })
+  } else {
+    return []
+  }
 })
 
-useWindowEvent('keypress', (e) => {
+function componentPos(index: number) {
+  return index * ITEM_SIZE
+}
+
+function componentAtY(pos: number) {
+  return Math.floor(pos / ITEM_SIZE)
+}
+
+function componentStyle(index: number) {
+  return { transform: `translateY(${componentPos(index)}px)` }
+}
+
+function componentColor(component: Component): string {
+  return componentStore.groups[component.group].color
+}
+
+// === Highlight ===
+
+const selected = ref(-1)
```

Comment:
I don't like the use of `-1` as an initialization value. It's usually better to use a nullable types when there is no valid initial state. Though in this case it's very convenient to always have a number, so I'd say keep it for simplicity sake. I would simply change the initial value to `0`, to not give anyone any ideas that there might be a special value of any kind.
```suggestion
const selected = ref(0)
```

Reply:
But I want to have no entry selected by default when Component Browser appears. Eventually, we will auto-select item after opening only on some occasions after reloading list.

Reply:
it's standard, but not necessarily a good idea - i think the convention came from java where of course your numbers cannot be `null`. in rust though i'd imagine it *should* be handled by an `Option` too

as for html itself - there is a lot of legacy spec features in html that even whatwg probably don't like nowadays (and a lot of inconsistency too)

but yeah i think `null` isn't a bad idea